### DISCUSSION TITLE
[feature request] make `transform` support async function

### DISCUSSION THREAD

**Discussion by harkor** [storybookjs/storybook#24236] [Ideas]
### Describe the bug

Since prettier 3.0.0, "prettier.format" return a Promise<string>

So in rendered code in stories, the code shown is "[object Promise]"

In Stackblitz, u can see the right html formatting inside console.

![image](https://github.com/storybookjs/storybook/assets/1044182/e09740d3-085f-4665-b705-cf65622fdc02)

```
# line 17 in .storybook/preview.js
    docs: {
      source: {
        transform: async (src, context) => {
          if (context.component) {
            const html = await prettier.format(decode(renderToStaticMarkup(context.originalStoryFn(context.args, context))), {
              parser: 'html',
              htmlWhitespaceSensitivity: 'ignore',
              plugins: [HTMLParser],
            });
            console.log(html);
            return html;
          }
        },
      },
    },
```

### To Reproduce

Stackblitz: https://stackblitz.com/github/harkor/storybook-prettier
Repo: https://github.com/harkor/storybook-prettier

### System

```shell
System:
    OS: Linux 5.15 Debian GNU/Linux 11 (bullseye) 11 (bullseye)
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i7-12700H
  Binaries:
    Node: 18.12.0 - ~/.nvm/versions/node/v18.12.0/bin/node
    Yarn: 3.6.1 - ~/.nvm/versions/node/v18.12.0/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.0/bin/npm
  Browsers:
    Chrome: 115.0.5790.170
  npmPackages:
    @storybook/addon-essentials: ^7.0.24 => 7.0.24 
    @storybook/addon-interactions: ^7.0.24 => 7.0.24 
    @storybook/addon-links: ^7.0.24 => 7.0.24 
    @storybook/addon-onboarding: ^1.0.8 => 1.0.8 
    @storybook/blocks: ^7.0.24 => 7.0.24 
    @storybook/react: ^7.0.24 => 7.0.24 
    @storybook/react-webpack5: ^7.0.24 => 7.0.24 
    @storybook/testing-library: ^0.2.0 => 0.2.0
```

**Comment by harkor** [comment]
Any clue ?

**Comment by MrErHu** [comment]
![image](https://github.com/storybookjs/storybook/assets/16449143/d0510863-ebd1-464f-8ba5-65791741daae)
not support async function

**Comment by vanessayuenn** [comment]
hi @harkor, thanks for opening this issue. @MrErHu is right -- `transform` currently only supports synchronous function. The feature request of supporting async is a reasonable one but unfortunately not on our immediate roadmap at the moment.

**Comment by omarkawach** [comment]
Commenting to say I would also be interested in this feature. Same use case with the prettier API

**Reply by rajsite** [reply]
We switched to [js-beautify](https://www.npmjs.com/package/js-beautify) with the html formatter configured as `wrap_attributes: 'force-expand-multiline'` due to the same issue. [Switch to beautify example PR](https://github.com/ni/nimble/pull/1918/files).

Would be nice to be able to use prettier so the snippets end up the same as in source usage but better than just being broken.  Would love for async support on transform to make the roadmap!

**Comment by seahindeniz** [comment]
Possibly duplicate of https://github.com/storybookjs/storybook/discussions/23703?

**Comment by valentinpalkovic** [comment]
Fixed by https://github.com/storybookjs/storybook/pull/30426

**Reply by c10b10** [reply]
Has this been released? When I return a promise from transform, clicking "Show source"  seems to not be doing anything.

**Reply by valentinpalkovic** [reply]
Can you please open an issue with a reproduction?

**Reply by valentinpalkovic** [reply]
And yes, released in Storybook 9.0

**Reply by c10b10** [reply]
Done: https://github.com/storybookjs/storybook/issues/32560
